plugins/talk-plugin-slack-notifications/server/hooks.js

```diff
@@ -0,0 +1,44 @@
+const fetch = require('node-fetch');
+const {SLACK_WEBHOOK_URL, SLACK_WEBHOOK_TIMEOUT} = require('./config');
+const debug = require('debug')('talk:plugin:slack-notifications');
+
+// We don't add the hooks during _test_ as the Slack API is not available.
+if (process.env.NODE_ENV === 'test') {
+  return null;
+}
+
+module.exports = {
+  RootMutation: {
+    createComment: {
+      async post(_, {input}, context, _info, result) {
+        debug(`Posting notification to Slack webhook: ${SLACK_WEBHOOK_URL}`);
```

Comment:
What I would suggest doing here is placing the call to Slack into the next tick, this will ensure that the operation will not block the comment creation process:

```js
process.nextTick(() => {
  // Your slack code goes here, don't return `result` from this function.
});

// Return it here instead.
return result;
```

Reply:
That's a pretty good idea, thanks! :+1:

Just one question to cement my understanding of the timing of events here. If you look at my 2nd commit, the callback to process.nextTick is marked `async`, so I think the order of execution will be:

post-resolve hook starts => schedules async callback for next tick of event loop => returns `result`, comment is created => makes request to Slack => until response arrives, Node processes other events / sits idle => reads response (again non-blocking)

Is this correct?

Reply:
Almost correct!

post-resolve hook starts => schedules async callback for next tick of event loop => returns result, comment is created => makes request to Slack => schedules processing of slack response when received, do other things, slack response came back => log ok!

So when the fetch goes out, it's now an I/O bound event, because it's a http request, so the system will switch over to doing things while it's waiting for a response.

Reply:
Makes sense. Probably uses kernel I/O events under the hood thanks to libuv (have briefly read about it in the past). :slightly_smiling_face: